### PR TITLE
document how to invoke system programs

### LINKED ISSUE #1759: LD_LIBRARY_PATH mod causes trouble for forked system applications
See there https://github.com/borgbackup/borg/pull/538 and there https://github.com/borgbackup/borg/issues/514.

Not sure there can be done anything against that issue in pyinstaller except maybe documenting that there is a problem and offering some easy way for the application to find out whether it is executing under pyinstaller and what was added to LD_LIBRARY_PATH by pyinstaller (it is not said that this path is otherwise empty) without it having to make guesses.

BTW, at the place where this gets done in pyinstaller, there is an interesting comment about that this would be troublesome on OS X. Well, seems it can also cause trouble on other POSIX OSes (as seen above).

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
fixes #1759.

### CODE REVIEW

File: doc/runtime-information.rst [hunk 1]
```diff
@@ -101,6 +101,36 @@ symbolic link::
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This environment variable is used to discover libraries, it is the library
+search path - on Linux and *BSD `LD_LIBRARY_PATH` is used, on AIX it is
+`LIBPATH`.
+
+PyInstaller saves the original value to `*_ORIG`, then modifies the search
+path so that the bundled libraries are found first by the bundled code.
+
+But if your code executes a system program, you often do not want that this
```
Comment (htgoebel) [inline doc/runtime-information.rst]: Many thanks for this doc update :-)

I would put this paragraph at the start of the section, so those who don't have this use case would not need to read further.

Comment (ThomasWaldmann) [inline doc/runtime-information.rst]: As it has backreferences that are needed to understand, this doesn't work.

File: doc/runtime-information.rst [hunk 2]
```diff
@@ -101,6 +101,36 @@ symbolic link::
...
+::
+
+    env = dict(os.environ)  # make a copy of the environment
+    lp_key = 'LD_LIBRARY_PATH'  # for Linux and *BSD.
+    lp_orig = env.get(lp_key + '_ORIG')  # pyinstaller >= 20160820 has this
+    if lp_orig is not None:
+        env[lp_key] = lp_orig  # restore the original, unmodified value
+    else:
+        env.pop(lp_key, None)  # last resort: remove the env var
+    p = Popen(system_cmd, ..., env=env)  # create the process
```
Comment (nchammas) [inline doc/runtime-information.rst]: @ThomasWaldmann - Isn't this equivalent to the code you have above?

```python
env = dict(os.environ)
env['LD_LIBRARY_PATH'] = env.get('LD_LIBRARY_PATH_ORIG', '')
p = Popen(system_cmd, ..., env=env)
```

Or must `LD_LIBRARY_PATH` be explicitly unset (as opposed to set to an empty string) when `LD_LIBRARY_PATH_ORIG` is not defined?

Comment (ThomasWaldmann) [inline doc/runtime-information.rst]: What you do in the "None case" is to create a env entry with value None. No idea whether that works.

Also, I have the lp_key value variable as for OS X you need a different value.

Comment (nchammas) [inline doc/runtime-information.rst]: In my snippet, if `LD_LIBRARY_PATH_ORIG` is not defined, then `LD_LIBRARY_PATH` will be set to the empty string, not `None`.

Anyway, I guess I should setup a Linux VM to test this out and see what happens.

**Pull request merged**